File: 53. Maximum Subarray/main.cpp
#include <vector>

using std::vector;

class Solution {
public:
    int maxSubArray(vector<int> &nums) {
        int max_sum = nums[0];
        int sum = nums[0];

        for (size_t i = 1; i < nums.size(); i++) {
            if (sum < 0) {
                sum = nums[i];
            } else {
                sum += nums[i];
            }

            if (max_sum < sum) {
                max_sum = sum;
            }
        }

        return max_sum;
    }
};

class Solution {
public:
    int maxSubArray(vector<int> &nums) {
        int max_sum = std::numeric_limits<int>::min();
        int sum = 0;

        for (const auto &num : nums) {
            sum = std::max(sum + num, num);
            max_sum = std::max(max_sum, sum);
        }

        return max_sum;
    }
};

class Solution {
public:
    int maxSubArray(vector<int> &nums) {
        return maxSubArrayHelper(nums, 0, nums.size() - 1);
    }

private:
    int maxSubArrayHelper(vector<int> &nums, int left, int right) {
        if (left == right) {
            return nums[left];
        }

        int mid = (left + right) / 2;
        int left_max = maxSubArrayHelper(nums, left, mid);
        int right_max = maxSubArrayHelper(nums, mid + 1, right);
        int crossing_max = maxCrossingSubArray(nums, left, mid, right);

        return std::max(left_max, std::max(right_max, crossing_max));
    }

    int maxCrossingSubArray(vector<int> &nums, int left, int mid, int right) {
        int left_sum = std::numeric_limits<int>::min();
        int sum = 0;

        for (int i = mid; i > left  - 1; i--) {
            sum += nums[i];
            left_sum = std::max(left_sum, sum);
        }

        int right_sum = std::numeric_limits<int>::min();
        sum = 0;

        for (int i = mid + 1; i <= right; i++) {
            sum += nums[i];
            right_sum = std::max(right_sum, sum);
        }

        return left_sum + right_sum;
    }
};

File: 53. Maximum Subarray/main.py
import sys
from typing import List


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        max_sum = -sys.maxsize - 1
        sum = 0

        for num in nums:
            sum = max(sum + num, num)
            max_sum = max(max_sum, sum)

        return max_sum


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        max_sum = nums[0]
        sum = nums[0]

        for i in range(1, len(nums)):
            if sum < 0:
                sum = nums[i]
            else:
                sum += nums[i]

            max_sum = max(max_sum, sum)

        return max_sum


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        return self.maxSubArrayHelper(nums, 0, len(nums) - 1)

    def maxSubArrayHelper(self, nums: List[int], left: int, right: int) -> int:
        if left == right:
            return nums[left]

        mid = (left + right) // 2
        left_max = self.maxSubArrayHelper(nums, left, mid)
        right_max = self.maxSubArrayHelper(nums, mid + 1, right)
        cross_max = self.maxCrossingSubArray(nums, left, mid, right)

        return max({left_max, right_max, cross_max})

    def maxCrossingSubArray(
        self, nums: List[int], left: int, mid: int, right: int
    ) -> int:
        left_sum = -sys.maxsize - 1
        sum = 0

        for i in range(mid, left - 1, -1):
            sum += nums[i]
            left_sum = max(left_sum, sum)

        right_sum = -sys.maxsize - 1
        sum = 0
        for i in range(mid + 1, right + 1):
            sum += nums[i]
            right_sum = max(right_sum, sum)

        return left_sum + right_sum


class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        return self.maxSubArrayHelper(nums, 0, len(nums) - 1)

    def maxSubArrayHelper(self, nums: List[int], left: int, right: int) -> int:
        if left == right:
            return nums[left]

        mid = (left + right) // 2
        left_max = self.maxSubArrayHelper(nums, left, mid)
        right_max = self.maxSubArrayHelper(nums, mid + 1, right)
        cross_max = self.maxCrossingSubArray(nums, left, mid, right)

        return max(left_max, right_max, cross_max)

    def maxCrossingSubArray(
        self, nums: List[int], left: int, mid: int, right: int
    ) -> int:
        left_sum = -sys.maxsize - 1
        sum = 0

        for num in nums[left : mid + 1][::-1]:
            sum += num
            left_sum = max(left_sum, sum)

        right_sum = -sys.maxsize - 1
        sum = 0
        for num in nums[mid + 1 : right + 1]:
            sum += num
            right_sum = max(right_sum, sum)

        return left_sum + right_sum
